adapter/pkg/xds/xds_server.go

```diff
@@ -274,3 +279,132 @@ func updateXdsCache(label string, endpoints []types.Resource, clusters []types.R
 	envoyUpdateVersionMap[label] = version
 	logger.LoggerMgw.Infof("New cache update for the label: " + label + " version: " + fmt.Sprint(version))
 }
+
+//different go routines could update XDS at the same time. To avoid this we use a mutex and lock
+func updateXdsCacheWithLock(label string, endpoints []types.Resource, clusters []types.Resource, routes []types.Resource,
+	listeners []types.Resource) {
+	mutexForXdsUpdate.Lock()
+	defer mutexForXdsUpdate.Unlock()
+	updateXdsCache(label, endpoints, clusters, routes, listeners)
+}
+
+func startConsulServiceDiscovery() {
+	//label := "default"
+	for apiKey, clusterList := range openAPIClustersMap {
+		//logger.LoggerXds.Debugln("API key for consul service discovery: ", apiKey)
+		//logger.LoggerXds.Info(svcdiscovery.ClusterConsulKeyMap)
+		for _, cluster := range clusterList {
+			logger.LoggerXds.Info(cluster.Name)
+			if consulSyntax, ok := svcdiscovery.ClusterConsulKeyMap[cluster.Name]; ok {
+				svcdiscovery.InitConsul() //initialize consul client and load certs
+				query, errConSyn := svcdiscovery.ParseQueryString(consulSyntax)
+				if errConSyn != nil {
+					logger.LoggerXds.Error("consul syntax parse error ", errConSyn)
+					return
+				}
+				logger.LoggerXds.Info(query)
+				go getServiceDiscoveryData(query, cluster.Name, apiKey)
+			}
+		}
+
+	}
+
+}
+
+func getServiceDiscoveryData(query svcdiscovery.Query, clusterName string, apiKey string) {
+
+	doneChan := make(chan bool)
+	svcdiscovery.ClusterConsulDoneChanMap[clusterName] = doneChan
+	resultChan := svcdiscovery.ConsulClientInstance.Poll(query, doneChan)
+	for {
+		select {
+		case queryResultsList, ok := <-resultChan:
+			if !ok { //ok==false --> result chan is closed
+				logger.LoggerXds.Info("closed the result channel for cluster name: ", clusterName)
+				return
+			}
+			//logger.LoggerXds.Info(queryResultsList)
+
+			//logger.LoggerXds.Info("map: ", svcdiscovery.ClusterConsulResultMap)
```

Comment:
Remove commented code

Reply:
Fix in other places in this file as well.

Reply:
fixed 5d85993